### ISSUE TITLE
Advice: Is there a way of sampling a % of spans

### ISSUE THREAD

**Issue by amccague** [open-telemetry/opentelemetry-collector-contrib#36876]
### Component(s)

pkg/ottl

### Describe the issue you're reporting

I have some noisy spans that I want to filter. But equally I want to know if they're still being produced and from where.

Ideally I'd implement a filter with a percentage, so 1% of the spans still get through. Though I can't see a rand function or something like modulo, though I could possibly craft it using divide. In the absence of rand I could try to use something like the last digit component as my random source.

This doesn't seem like a novel idea though so potentially it's a solved problem and my searching skills aren't up to scatch.

These are leaf spans so it wouldn't impact any children.
I don't think this is worthy use of the tail sampling filter either.

**Comment by mkromkamp** [comment]
Bot the [tailsampling](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor) and the [probabilistic sampling processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/probabilisticsamplerprocessor) support this scenario.

There is a small explanation when to choose which [over here](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/tailsamplingprocessor).

Hope that helps.

**Comment by amccague** [comment]
Hi @mkromkamp - I have reviewed those two processors and to be honest I can't quite determine from the configuration examples in the readme how to achieve this. They appear to sample _traces_ first and foremost.

I can't see the specific configurations to target a span by say attribute value or name. Could you be more specific about the configuration or samplers you had in mind?

Thanks

**Comment by mkromkamp** [comment]
> Hi @mkromkamp - I have reviewed those two processors and to be honest I can't quite determine from the configuration examples in the readme how to achieve this. They appear to sample _traces_ first and foremost.
> 
> I can't see the specific configurations to target a span by say attribute value or name. Could you be more specific about the configuration or samplers you had in mind?
> 
> Thanks

Yeah sure. I might need a concrete example. But _# Rule 2: low sampling for readiness/liveness probes_ sounds roughly like what you are potentially looking for.

It keeps 10% (0.1) percent of traces for the `/live` and `/ready` endpoints for services with the name `service-1`, `service-2`, and `service-3`

**Comment by amccague** [comment]
@mkromkamp 

This would as I understand it sample the whole (sub)trace for that route. Per my original request I want to probability sample a particular _span_, I want to keep the rest of the spans for that given route.

**Comment by mkromkamp** [comment]
> @mkromkamp
> 
> This would as I understand it sample the whole (sub)trace for that route. Per my original request I want to probability sample a particular _span_, I want to keep the rest of the spans for that given route.

I see, that wasn't clear to me from the original post. And is indeed not possible with this approach 🙂

**Comment by jmacd** [comment]
I take it you are interested in what the probabilistic sampler does (which is to sample individual spans), but that you need to conditionalize the logic by span name or other attribute. The tail sampler has the type of condition you require, but samples whole traces.

While not exactly what you're asking for, there is work at the OTel specification level for composite samplers which includes a rule-based approach combined with consistent probability sampling. It would be natural to extend the probabilistic sampler processor to support such features. See https://github.com/open-telemetry/opentelemetry-specification/pull/4321

**Comment by jmacd** [comment]
@amccague I agree that your use-case is not directly handled by `probabilisticsamplingprocessor` even though it almost would, or could, with more development. The term I use to describe what you're interested in is "intermediate span sampling". The missing piece is a rule engine to decide which probability to use based on span attributes / latency / etc, the way `tailsampler` does across traces, but limited to a span or log record (i.e., the scope of `probabilisticsamplingprocessor`). There is connected work at the specification level on this topic here: https://github.com/open-telemetry/opentelemetry-specification/pull/4321
and here: https://github.com/open-telemetry/opentelemetry-specification/pull/4166

**Comment by jmacd** [comment]
@amccague In the issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38468, there are some examples of how the logs priority sampling mechanism can allow OTTL transforms to configure sampling policies for log records. The same would work for spans, except that the span sampling logic is different.

Given how OTTL makes this sampling component composable, I think it should do the same for traces as it has always done for logs. 

Here is the example for logs.

```
processors:
  transform:
    log_statements:
    - context: log
      statements:
        - set(attributes["sampling_uuid"], UUID()) 
        - set(attributes["sampling_priority"], 10)
          where IsMatch(log.body, "noisy")

  attributes:
    actions:
    - key: sampling_priority
      value: 100
      action: upsert

  probabilistic_sampler:
    sampling_percentage: 100
    sampling_priority: sampling_priority
    attribute_source: record
    from_attribute: sampling_uuid
```

This means fixing the incompatibility noted in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/probabilisticsamplerprocessor/README.md#sampling-priority.
